drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -48,90 +67,258 @@ static int cmd_reg_read(const struct device *dev, uint8_t reg_addr, int16_t *val
 
-	reg_addr = BQ274XX_CMD_CONTROL_HIGH;
-	i2c_data = (uint8_t)((subcommand >> 8) & 0x00FF);
+	i2c_data = BQ274XX_EXT_BLKDAT_START + offset;
 
-	ret = i2c_reg_write_byte_dt(&config->i2c, reg_addr, i2c_data);
+	ret = i2c_burst_read_dt(&config->i2c, i2c_data, data, bytes);
 	if (ret < 0) {
-		LOG_ERR("Failed to write into control high register");
+		LOG_ERR("Failed to read block");
 		return -EIO;
 	}
 
+	k_msleep(BQ274XX_SUBCLASS_DELAY_MS);
+
 	return 0;
 }
 
-static int cmd_reg_write(const struct device *dev, uint8_t command, uint8_t data)
+static int bq274xx_get_device_type(const struct device *dev, uint16_t *val)
 {
-	const struct bq274xx_config *config = dev->config;
-	uint8_t i2c_data, reg_addr;
-	int ret = 0;
+	int ret;
 
-	reg_addr = command;
-	i2c_data = data;
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_DEVICE_TYPE);
+	if (ret < 0) {
+		LOG_ERR("Unable to write control register");
+		return -EIO;
+	}
 
-	ret = i2c_reg_write_byte_dt(&config->i2c, reg_addr, i2c_data);
+	ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_CONTROL_LOW, val);
 	if (ret < 0) {
-		LOG_ERR("Failed to write into control register");
+		LOG_ERR("Unable to read register");
 		return -EIO;
 	}
 
 	return 0;
 }
 
-static int read_data_block(const struct device *dev, uint8_t offset,
-			   uint8_t *data, uint8_t bytes)
+static int bq274xx_gauge_configure(const struct device *dev)
 {
-	const struct bq274xx_config *config = dev->config;
-	uint8_t i2c_data;
-	int ret = 0;
+	const struct bq274xx_config *const config = dev->config;
+	struct bq274xx_data *data = dev->data;
+	struct bq274xx_regs *regs = data->regs;
+	int ret;
+	uint8_t tmp_checksum, checksum_old, checksum_new;
+	uint16_t flags, designenergy_mwh, taperrate, reg_val;
+	uint8_t block[BQ27XXX_DM_SZ];
+	uint8_t try;
 
-	i2c_data = BQ274XX_EXT_BLKDAT_START + offset;
+	designenergy_mwh = (uint32_t)config->design_capacity * 3700 / 1000;
+	taperrate = config->design_capacity * 10 / config->taper_current;
 
-	ret = i2c_burst_read_dt(&config->i2c, i2c_data, data, bytes);
+	/* Unseal the battery control register */
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_A);
 	if (ret < 0) {
-		LOG_ERR("Failed to read block");
+		LOG_ERR("Unable to unseal the battery");
 		return -EIO;
 	}
 
-	k_msleep(BQ274XX_SUBCLASS_DELAY);
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_B);
+	if (ret < 0) {
+		LOG_ERR("Unable to unseal the battery");
+		return -EIO;
+	}
 
-	return 0;
-}
+	/* Send CFG_UPDATE */
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_SET_CFGUPDATE);
+	if (ret < 0) {
+		LOG_ERR("Unable to set CFGUpdate");
+		return -EIO;
+	}
 
-static int get_device_type(const struct device *dev, uint16_t *val)
-{
-	int ret;
+	/* Step to place the Gauge into CONFIG UPDATE Mode */
+	try = 100;
+	do {
+		ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_FLAGS, &flags);
+		if (ret < 0) {
+			LOG_ERR("Unable to read flags");
+			return -EIO;
+		}
+
+		if (!(flags & BQ27XXX_FLAG_CFGUP)) {
+			k_msleep(BQ274XX_SUBCLASS_DELAY_MS * 10);
```

Comment:
It would be better if `BQ274XX_SUBCLASS_DELAY_MS * 10` had it's own `#define`.

Reply:
Done.